python intermediate/hero_game/classes6.py

```diff
@@ -0,0 +1,193 @@
+
+
+import random
+
+
+N = 25
+M = 25
+K = 2
+PROPORTION = 1
+
+
+class Hero:
+    def __init__(self, name: str) -> object:
+        self.name = name
+        self.level = 1
+        self.damage = 2
+        self.hp = 10
+        self.coins = 0
+        self.full_hp = 10 #in order to remember the full hp
+        self.defend = False #is the defend option is on
+
+    
+    def heal(self) -> None:
+        """
+        this function heals the character by N percent
+        """
+        self.hp = self.hp * (N/100 + 1)        
+
+    
+    def level_up(self) -> bool: 
+        """
+        this function level up a character 
+        """
+        if self.level + 1 * K < self.coins:
+            self.level += 1
+            self.damage = self.damage * (M/100 + 1)
+            self.full_hp = self.full_hp * (M/100 + 1)
+            self.hp = self.full_hp
+            return True
+        
+        return False
+
+
+    def attack(self, monster) -> None:
+        """
+        attack monster function
+        """
+        monster.reduce_health(self)
+        
+        if monster.health <= 0:
+            self.coins += self.level
+
+    
+    def is_defend(self) -> None:
+        """
+        activates the defend function
+        """
+        self.defend = True
+
+    
+    def reduce_health(self, monster) -> float:
+        """
+        the reduce health function
+        """
+        current_health = self.hp
+        diff = current_health - (self.hp * 1.8) # help the logics of the game
+
+        if self.defend:
+            self.hp = self.hp * 1.8
+
+        self.hp -= monster.damage
+
+        if self.hp < 0:
+            self.hp = 0
+
+        if self.defend:
+            if diff > monster.damage:
```

Comment:
because diff is always negative, and the damage cant be negative, this is never true